v0/destinations/blueshift/transform.js

```diff
@@ -0,0 +1,159 @@
+  }
+  const response = defaultRequestConfig();
+
+  if (destination.Config.datacenterEU) {
+    response.endpoint = `${BASE_URL_EU}/api/v1/event`;
+  } else {
+    response.endpoint = `${BASE_URL}/api/v1/event`;
+  }
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.headers = {
+    "Content-Type": "application/json"
+  };
+  response.body.JSON = payload;
+  return response;
+};
+
+const identifyResponseBuilder = async (message, category, destination) => {
+  const payload = constructPayload(message, MAPPING_CONFIG[category.name]);
+
+  if (!payload) {
+    // fail-safety for developer error
+    throw new CustomError(ErrorMessage.FailedToConstructPayload, 400);
+  }
+  const response = defaultRequestConfig();
+
+  if (destination.Config.datacenterEU) {
+    response.endpoint = `${BASE_URL_EU}/api/v1/customers`;
+  } else {
+    response.endpoint = `${BASE_URL}/api/v1/customers`;
+  }
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.headers = {
+    "Content-Type": "application/json"
+  };
+  response.body.JSON = payload;
+  return response;
+};
+
+const groupResponseBuilder = async (message, category, destination) => {
+  const payload = constructPayload(message, MAPPING_CONFIG[category.name]);
+
+  if (!payload) {
+    // fail-safety for developer error
+    throw new CustomError(ErrorMessage.FailedToConstructPayload, 400);
+  }
+  const response = defaultRequestConfig();
+
+  const customer = removeUndefinedAndNullValues({
+    id: getFieldValueFromMessage(message, "userId"),
+    email: getFieldValueFromMessage(message, "email")
+  });
+  if (!customer.id && !customer.email) {
+    throw new CustomError(
+      "[Blueshift] customer_id or email is required to identify customer.",
+      400
+    );
+  }
+  if (customer.id) {
+    payload.identifier_key = "customer_id";
+    payload.identifier_value = customer.id;
+  } else {
+    payload.identifier_key = "email";
+    payload.identifier_value = customer.email;
+  }
+
+  if (destination.Config.datacenterEU) {
+    response.endpoint = `${BASE_URL_EU}/api/v1/custom_user_lists/add_user_to_list/${payload.list_id}`;
+  } else {
+    response.endpoint = `${BASE_URL}/api/v1/custom_user_lists/add_user_to_list/${payload.list_id}`;
+  }
+
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.headers = {
+    "Content-Type": "application/json"
+  };
+  response.body.JSON = payload;
+  return response;
+};
+
+const process = async event => {
+  const { message, destination } = event;
+  if (!message.type) {
+    throw new CustomError(
+      "Message Type is not present. Aborting message.",
+      400
+    );
+  }
+
+  const messageType = message.type.toLowerCase();
+  const category = CONFIG_CATEGORIES[message.type.toUpperCase()];
+
+  let response;
+  switch (messageType) {
+    case EventType.TRACK:
+      response = await trackResponseBuilder(message, category, destination);
```

Comment:
Add API key validation for each of these types.

Reply:
Rather than validating at every function validate it on process